packages/application/src/shell.ts

```diff
@@ -289,4 +491,339 @@ namespace Private {
     private _items = new Array<Private.IRankItem>();
     private _panel = new Panel();
   }
+
+  /**
+   * A class which manages a side bar that can show at most one widget at a time.
+   */
+  export class SideBarHandler {
+    /**
+     * Construct a new side bar handler.
+     */
+    constructor(area: 'left' | 'right') {
+      this._area = area;
+      this._stackedPanel = new StackedPanel();
+      this._stackedPanel.hide();
+      this._current = null;
+      this._lastCurrent = null;
+      this._stackedPanel.widgetRemoved.connect(this._onWidgetRemoved, this);
+      this._sideBarPalette = null;
+      this._menuEntryLabel = `${area[0].toUpperCase()}${area.slice(1)} Sidebar`;
+    }
+
+    get current(): Widget | null {
+      return (
+        this._current ||
+        this._lastCurrent ||
+        (this._items.length > 0 ? this._items[0].widget : null)
+      );
+    }
+
+    /**
+     * Whether the panel is visible
+     */
+    get isVisible(): boolean {
+      return this._stackedPanel.isVisible;
+    }
+
+    /**
+     * Get the stacked panel managed by the handler
+     */
+    get stackedPanel(): StackedPanel {
+      return this._stackedPanel;
+    }
+
+    /**
+     * Signal fires when the stacked panel changes
+     */
+    get updated(): ISignal<SideBarHandler, void> {
+      return this._updated;
+    }
+
+    /**
+     * Add a function to update the menus with the side bar widgets.
+     * @param updateMenu - the function to call to update the menu.
+     */
+    addUpdateMenuFn(updateMenu: SideBarPanel.UpdateSideBarMenuFn): void {
+      this._updateMenu = updateMenu;
+      this.updateMenu();
+    }
+
+    /**
+     * Associate palette entries to the sidebar, and update it with the current widgets.
+     */
+    createPaletteEntry(options: SideBarPaletteOption): void {
```

Comment:
Wondering if it would also be possible to move the command palette related logic to the plugin level as well?

It looks like the palette and the shell shouldn't be too coupled to allow for greater flexibility in how the plugins are composed (or if some are disabled by end users). There could be applications without palettes or menus and having references to these in the shell implementation feels odd.

So leaving the shell handle and track the widgets with a stacked panel. And the plugin build the appropriate menu and palette entries.

Reply:
Yes, maybe we could add a module in the extension with this `sideBarPalette`, and pass a palette update function to the Shell, like for the Menu.